pilot/pkg/proxy/envoy/v2/discovery.go

```diff
@@ -158,6 +162,19 @@ func (s *DiscoveryServer) Register(rpcs *grpc.Server) {
 	ads.RegisterAggregatedDiscoveryServiceServer(rpcs, s)
 }
 
+// OnServerReady is called when caches have been synced so that it can process requests.
+func (s *DiscoveryServer) OnServerReady() {
+	s.updateMutex.Lock()
+	s.serverReady = true
```

Comment:
Should we just recreate the context here?

Reply:
Yeah, that  seems to be  an option - We can add `RefreshContext` method in `PushContext` that just reloads every thing without checking if init is done or not, and we can call it once caches have been synced and before we declare our self ready.

Only downside of it is pilot's local proxy would get incomplete config in the first call but would immediately get refreshed because of config change and since it is any way starting afresh, it should be OK.

If this is an option, I can make that change.

@hzxuzhonghu What do you think about this?

Reply:
Sorry, i could not catch up, what do you mean by `context ` here?

Reply:
`PushContext` -  recreate the PushContext once caches have been synced

Reply:
Let me think about

Reply:
We donot need refresh pushContext here, when an xds stream established, the first thing to do is initializing pushContext.

https://github.com/istio/istio/pull/20296/files#diff-14f4a8a239bf5ba82053cf6606daace5R210

Reply:
That is correct. But the question @howardjohn  is asking is refresh the context here and  remove this https://github.com/istio/istio/blob/master/pilot/pkg/proxy/envoy/v2/ads.go#L194 to reject connections - That way the pilot's local Envoy would not be rejected (and get latest config after config update) and the other proxies that connect to will have updated context automatically.

Reply:
https://github.com/istio/istio/pull/20323 - This is what he is suggesting. PTAL if it is OK or it causes any issues

Reply:
Got it.  Then we can do this:  trigger full push to all connected clients  here.

Reply:
But there is one concern: there may be race condition  between the first xds request/push and the new push here.

Reply:
There wont be any new clients here except for local pilot proxy because Pilot is not ready it self  here, it will only be ready after this line. Regarding pilot's proxy, I guess it would get updates because of regular configupdate push and since it is any way starting now, It should not be an issue.

Reply:
make sense